### PR TITLE
ScaledReordered interface

### PR DESCRIPTION
**Pull request by fritzgoebel**
This PR adds an interface for wrapping an inner operator (like a direct or iterative solver) into a (for now symmetrical, e.g. RCM) reordering via a reordering factory and diagonal scaling via one row and one column scaling diagonal matrix.

The `ScaledReordered` LinOp applies the given scaling and reordering to the system matrix and generates the inner operator on the modified problem. It manages three vectors, a right-hand side and a solution vector for the inner operator, and an intermediate vector used for the permutations.

With a permutation matrix `P`, a row scaling matrix `R`, and a column matrix `C`, the system matrix and vectors are scaled and permuted such that instead of the original system of equations
`A*x = b`
the inner operator solves 
`P*R*A*C*P^T*y = P*R*b`.
Hence, when generating a `ScaledReordered` LinOp, the new system matrix `P*R*A*C*P^T` is computed. When it is applied to vectors `b` and `x`, the right-hand side vector is modified to `P*R*b`, and the solution `y` to the inner system gives the solution `x` to the original system via `x = C*P^T*y`.
When using an inner operator that uses an initial guess, currently the initial guess for the inner system is computed by `y_0 = P*C^(-1)*x_0`, which is the same as using `x_0` for the system `P*R*A*x = P*R*b`. For the inverse diagonal scaling here I added an `inverse_apply` to the `matrix::Diagonal`. I still need to add tests for that.

Any feedback is welcome, especially on ideas regarding how to handle algorithms like RCM that generate a reordering, but also be able to handle MC64 that generates both a reordering and scaling coefficients and simpler equilibration techniques that only generate scaling coefficients.

### CODE REVIEW

**Comment by fritzgoebel** [comment]
format!

**Comment by fritzgoebel** [comment]
format!

**Review by yhmtsai — COMMENTED** [review]
I am thinking about removing the ValueType and IndexType from the ScaledReordered.
set_cache_to you can use create_with_config_of to create cache.
some requirement of ValueType / IndexType can be done by run<type-list>
Will it use ScaledReordered even when only need scale or reorder part?
Not a Scaled and Reordered class to handle them separately and ScaledReordered inherit from Scaled and Reordered.

File: include/ginkgo/core/base/reordered.hpp
Comment (upsj) [reply include/ginkgo/core/base/reordered.hpp:143]: #988 <---
for some LinOps, this is perfectly safe - in-place scaling is one of them.

File: include/ginkgo/core/base/reordered.hpp
Comment (upsj) [reply include/ginkgo/core/base/reordered.hpp:204]: `apply_uses_initial_guess()`?

File: include/ginkgo/core/base/reordered.hpp
Comment (yhmtsai) [reply include/ginkgo/core/base/reordered.hpp:143]: Yes, at least in my imagine about what compiler works. TBH, I wouldn't like to challenge the compiler did the same thing as imagine.

File: include/ginkgo/ginkgo.hpp
Comment (pratikvn) [inline include/ginkgo/ginkgo.hpp:65]: I would rename this to `scaled_reordering` ? Just `reordered` is a bit vague.

File: include/ginkgo/core/reorder/reordering_base.hpp
Comment (pratikvn) [inline include/ginkgo/core/reorder/reordering_base.hpp:67]: Is it necessary to have these getters here ?

**Review by pratikvn — APPROVED** [review]
Nice work. LGTM!

**Comment by yhmtsai** [comment]
AFAIR, combing scaled and reordered together allows some algorithms uses only one operation.
how do you do that when a new LinOp?

File: include/ginkgo/core/reorder/reordering_base.hpp
Comment (fritzgoebel) [reply include/ginkgo/core/reorder/reordering_base.hpp:67]: As the scaled reordered factory takes a `ReorderingBaseFactory`, I think these are necessary here rather than only in the algorithm implemenations.

File: include/ginkgo/core/base/reordered.hpp
Comment (fritzgoebel) [reply include/ginkgo/core/base/reordered.hpp:48]: Yes, before merging this we should agree on naming and placement.

**Comment by fritzgoebel** [comment]
> AFAIR, combing scaled and reordered together allows some algorithms uses only one operation. how do you do that when a new LinOp?

At the moment, not at all. So far the operations are not merged yet.

**Comment by yhmtsai** [comment]
will the merged operations be in this class or another class inherited from this class?
if there's another class, the clone will only give the clone of ScaledReordered class not itself.

File: include/ginkgo/core/reorder/reordering_base.hpp
Comment (thoasm) [inline include/ginkgo/core/reorder/reordering_base.hpp:66]: nit:
```suggestion
    virtual std::shared_ptr<const LinOp> get_permutation() const = 0;

    virtual std::shared_ptr<const LinOp> get_inverse_permutation()
```
Also note that this breaks the interface of RCM (as you had `std::shared_ptr<matrix::Permutation>` as the return type).

File: include/ginkgo/core/reorder/rcm.hpp
Comment (thoasm) [inline include/ginkgo/core/reorder/rcm.hpp:112]: This change is an actual interface break.
Previously, you could have called
```
rcm->get_permutation()->get_permutation_size();
```
However, now, you need to do it with a cast since you no longer know the actual type of the permutation:
```
as<matrix::Permutation>(rcm->get_permutation())->get_permutation_size();
```
Currently, this affects only RCM (as there is no other reordering).
Is it necessary to make the permutation type a LinOp (meaning are you expecting other permutation types)?

File: reference/test/base/reordered.cpp
Comment (thoasm) [inline reference/test/base/reordered.cpp:188]: nit  (for AAA separation):
```suggestion
        SR::build().with_row_scaling(this->diag).on(this->exec);
        
    auto scaled_reordered = scaled_reordered_fact->generate(this->mtx);
```

File: reference/test/base/reordered.cpp
Comment (thoasm) [inline reference/test/base/reordered.cpp:201]: same here:
```suggestion
        SR::build().with_col_scaling(this->diag).on(this->exec);
        
    auto scaled_reordered = scaled_reordered_fact->generate(this->mtx);
```

File: reference/test/base/reordered.cpp
Comment (thoasm) [inline reference/test/base/reordered.cpp:216]:
```suggestion
                                     .on(this->exec);
                                     
    auto scaled_reordered = scaled_reordered_fact->generate(this->mtx);
```

File: reference/test/base/reordered.cpp
Comment (thoasm) [inline reference/test/base/reordered.cpp:229]:
```suggestion
        SR::build().with_reordering(this->rcm_factory).on(this->exec);
        
    auto scaled_reordered = scaled_reordered_fact->generate(this->rcm_mtx);
```

File: reference/test/base/reordered.cpp
Comment (thoasm) [inline reference/test/base/reordered.cpp:264]: nit: for cleaner AAA separation:
```suggestion
    using SR = typename TestFixture::SR;
    auto scaled_reordered_fact = SR::build()
                                     .with_row_scaling(this->diag)
                                     .with_col_scaling(this->diag)
                                     .with_reordering(this->rcm_factory)
                                     .with_inner_operator(this->solver_factory)
                                     .on(this->exec);
    auto scaled_reordered = scaled_reordered_fact->generate(this->mtx);
    auto before_system_matrix = scaled_reordered->get_system_matrix();
    auto before_inner_operator = scaled_reordered->get_inner_operator();

    auto copied = SR::build().on(this->exec)->generate(this->rcm_mtx);
    copied->copy_from(scaled_reordered.get());

    ASSERT_EQ(before_system_matrix, copied->get_system_matrix());
    ASSERT_EQ(before_inner_operator, copied->get_inner_operator());
```

File: reference/test/base/reordered.cpp
Comment (thoasm) [inline reference/test/base/reordered.cpp:279]: same here
```suggestion
    auto scaled_reordered = scaled_reordered_fact->generate(this->mtx);
    auto before_system_matrix = scaled_reordered->get_system_matrix();
```

File: reference/test/base/reordered.cpp
Comment (thoasm) [inline reference/test/base/reordered.cpp:301]: and here:
```suggestion
    auto scaled_reordered = scaled_reordered_fact->generate(this->mtx);
    auto before_system_matrix = scaled_reordered->get_system_matrix();
```

File: include/ginkgo/core/base/reordered.hpp
Comment (thoasm) [inline include/ginkgo/core/base/reordered.hpp:165]: This requires your reordering to return a `matrix::Permutation`. However, this is not documented and not intuitive (your `ReorderingBaseFactory` needs `get_permutation()` to return a `matrix::Permutation`).
Is there a reason why you don't call the `apply` method on the permutation matrix? That would allow for more formats than just `matrix::Permutation` and would work a lot better with your new interface.

File: include/ginkgo/ginkgo.hpp
Comment (thoasm) [reply include/ginkgo/ginkgo.hpp:65]: I second that.

File: include/ginkgo/core/base/reordered.hpp
Comment (thoasm) [inline include/ginkgo/core/base/reordered.hpp:216]: Same here: if you want it to work for all reordering options, it should be a plain `LinOp` or a new data format (it should match the `get_permutation()` return type of your `ReorderingBase`).

**Review by thoasm — CHANGES_REQUESTED** [review]
There is an interface break at the moment (the only reason why I don't approve this PR).
Unfortunately, it will be quite difficult to find a solution that does not break the existing interface while being intuitive to use (for example, adding `get_permutation_2` would not break the interface, but be quite confusing for the user).
Additionally, the actual reordering inside the constructor of `ScaledReordered` needs a bit more work.

**Comment by fritzgoebel** [comment]
format!

File: include/ginkgo/core/base/reordered.hpp
Comment (fritzgoebel) [reply include/ginkgo/core/base/reordered.hpp]: I went with an array now, that works for all kinds of permutations and does not assume we have a permutation matrix.

File: include/ginkgo/core/reorder/scaled_reordered.hpp
Comment (thoasm) [inline include/ginkgo/core/reorder/scaled_reordered.hpp:107]: Requires a note that the system matrix must be of type `Permutable<index_type>` if a reordering is specified.

File: include/ginkgo/core/reorder/scaled_reordered.hpp
Comment (thoasm) [inline include/ginkgo/core/reorder/scaled_reordered.hpp:165]: Without a note, this `as<Permutable<index_type>>` might surprise some users.

**Review by thoasm — APPROVED** [review]
Thanks for the improvements. I have only minor requests, after addressing these, you can feel free and merge it.

**Review by upsj — APPROVED** [review]
LGTM for what we can currently change - I think there is still a lot of work to be done in the Permutation, Diagonal and ScaledPermutation area, but that needs to happen in Ginkgo 2.0

File: include/ginkgo/core/reorder/scaled_reordered.hpp
Comment (yhmtsai) [inline include/ginkgo/core/reorder/scaled_reordered.hpp:247]:
```suggestion
        std::unique_ptr<matrix::Dense<value_type>> inner_b{};
        std::unique_ptr<matrix::Dense<value_type>> inner_x{};
        std::unique_ptr<matrix::Dense<value_type>> intermediate{};
```
nit: these are data

File: include/ginkgo/core/reorder/scaled_reordered.hpp
Comment (yhmtsai) [inline include/ginkgo/core/reorder/scaled_reordered.hpp:174]: need to put the size, apply will check the size with b, x

File: reference/test/reorder/scaled_reordered.cpp
Comment (yhmtsai) [inline reference/test/reorder/scaled_reordered.cpp:312]: need a test for no inner_factory

**Review by yhmtsai — CHANGES_REQUESTED** [review]
LGTM in general. I have discussed with Fritz about the combining the permutation and scaling. It will change the implementation not a new class, so my above worry will not happen. 

I put it is request changes because it may give error when inner_operator is nullptr (generated Identity)

File: include/ginkgo/core/reorder/reordering_base.hpp
Comment (yhmtsai) [inline include/ginkgo/core/reorder/reordering_base.hpp:68]: you can also provide the setter for permutation_array in protected and the permutation data in private.
without implement it in each class

File: include/ginkgo/core/reorder/reordering_base.hpp
Comment (yhmtsai) [inline include/ginkgo/core/reorder/reordering_base.hpp:87]: public interface?

**Review by yhmtsai — APPROVED** [review]
LGTM in general. two concern: 1. public interface from using alias 2. move the get_permutation_array into ReorderBase.

File: include/ginkgo/core/reorder/reordering_base.hpp
Comment (fritzgoebel) [reply include/ginkgo/core/reorder/reordering_base.hpp:87]: I added a template argument to these two. But I think technically, the second one is still an interface break since in case someone used the `PolymorphicBase` template argument, this would no longer work...

Comment (thoasm) [reply include/ginkgo/core/reorder/reordering_base.hpp:87]: I'm fine if you break the interface at that point as I don't think a lot of people (my assumption is 0 outside of us) were using it. If it is compiled with `c++17`, the interface is no longer broken, so I would consider it as acceptable.

**Comment by fritzgoebel** [comment]
Depends on  #1171  now.

**Comment by fritzgoebel** [comment]
Ready to merge after #1171

**Pull request merged**
